### ISSUE TITLE
Replace ARPACK

### ISSUE THREAD

**Issue by gaborcsardi** [igraph/igraph#551]
Possible solutions to look at: 
- arpack-ng https://github.com/pv/arpack-ng
- feast http://www.ecs.umass.edu/~polizzi/feast/features.htm
- eigen http://eigen.tuxfamily.org/index.php?title=Main_Page 
- more ??? check http://www.grycap.upv.es/slepc/documentation/reports/str6.pdf

Cons:
- arpack-ng is the same codebase, just repackaged?
- feast does not have a non-symmetric solver
- eigen is C++ templates 
- eigen is MPL 2.0, which is somewhat problematic. (If we do not change anything in the source, then eigen files will be dual licensed, if we change something in it, the same happens, but care must be taken not to put any (L)GPL code in these files.) Actually, eigen might not even have sparse eigensolvers....

**Comment by gaborcsardi** [comment]
The ARPACK Fortran code in rARPACK seems to work better, so it is worth just trying to take their ARPACK version.

**Comment by gaborcsardi** [comment]
Also, http://www.math.uri.edu/~jbaglama/ has some Matlab code irlba is based on, so it is better to rewrite the original Matlab code into C. They also have a non-symmetric solver. The code is fairly involved, though.

**Comment by szhorvat** [comment]
Reasons for moving away from ARPACK: #544, #1641 .

**Comment by szhorvat** [comment]
FEAST now supports non-symmetric problems, and seems like a realistic alternative.

**Comment by ValZapod** [comment]
> seems like a realistic alternative

But it is part of Intel oneMKL, so "not allowed" on non-Intel. Also is not it closed source, is not it a problem?

**Comment by szhorvat** [comment]
The one I linked to appears to be BSD-licensed and presumably portable Fortran code.

**Comment by ntamas** [comment]
It seems promising, but FEAST does not seem to work on macOS (Apple M1) yet; I tried this in the `src/` folder:

```sh
$ make ARCH=x64 F90=gfortran MKL=no feast
gfortran -O3 -fopenmp -ffree-line-length-none -ffixed-line-length-none -cpp  -c /Users/tamas/Downloads/FEAST/4.0/src/kernel/dzfeast.f90 -o /Users/tamas/Downloads/FEAST/4.0/src/kernel/dzfeast.o
/Users/tamas/Downloads/FEAST/4.0/src/kernel/dzfeast.f90:917:82:

  650 |      call DGEMM('T','N',fpm(23),fpm(23),N,-DONE,work,N,q,N,DZERO,Aq,M0) ! projection
      |                                                                 2
......
  917 |      call DGEMM('N','N',N,fpm(25),fpm(23),DONE,work(1,1),N,Aq(1,fpm(24)),M0,DZERO,workc(1,fpm(24)),N)
      |                                                                                  1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (COMPLEX(8)/REAL(8)).
/Users/tamas/Downloads/FEAST/4.0/src/kernel/dzfeast.f90:944:87:

  839 |            call DLACPY( 'F', fpm(23), fpm(23),Bq, M0, Bqo, fpm(23) )
      |                                                      2
......
  944 |      if ((fpm(5)==1).and.(loop==0))  call DLACPY( 'F', N, fpm(25),work(1,fpm(24)) , N, workc(1,fpm(24)), N )
      |                                                                                       1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (COMPLEX(8)/REAL(8)).
make: *** [/Users/tamas/Downloads/FEAST/4.0/src/kernel/dzfeast.o] Error 1
```

The default `make` invocation tries to use `ifort` (which I don't have), it may be the case that it works with `ifort`. I'm not well-versed in Fortran so I'd rather not attempt debugging what's going on yet.

**Comment by szhorvat** [comment]
`ifort` is Intel Fortran, I believe.

I did wonder about compatibility with arm64, precisely because of the association with Intel, so I checked if MacPorts has arm64 binaries, and it does: https://ports.macports.org/port/feast/details/ They don't seem to use any obvious patches either: https://github.com/macports/macports-ports/blob/master/math/feast/Portfile

**Comment by szhorvat** [comment]
@ntamas MacPorts uses the gfortran option `-fallow-argument-mismatch`, then it works. I edited the Makefile to add this option, but there should be better ways.

**Comment by szhorvat** [comment]
Here's a recent paper comparing various eigensolvers:

https://n.ethz.ch/~gtzounas/pap/eigmethods.pdf
https://doi.org/10.3390/app10217592

Summary of features:

<img width="828" alt="image" src="https://user-images.githubusercontent.com/1212871/221870333-bfdcc060-041a-4029-a3eb-abb72a706314.png">

Based on this table, we should investigate SLEPc and z-PARES.

We absolutely need:

 - Non-symmetric problems, and these all have it
 - A reverse communication interface in order to use our own graph data structure directly. This is indicated as "RCI" in the table above.

**Comment by szhorvat** [comment]
SLEPc appears to be designed for massively parallel use, with MPI. It's unclear if it's also realistic for solving smaller problems without parallelization, and whether it can be compiled with MPI. If not, that would be a dealbreaker. Looking at how this is packaged in MacPorts, SLEPc depends on PETSc, which depends on MPI and BLAS/LAPACK, so there are not a lot of dependencies (apart from MPI), but it may still be a heavyweight, too large to vendor. It seems that this might be buildable without Fortran, but it's not completely clear.

**Comment by szhorvat** [comment]
[z-PARES](https://zpares.cs.tsukuba.ac.jp/) is also focused on parallelization with MPI, but it is stated on their homepage that serial operation is also supported. It is written in Fortran 90/95, so it cannot be f2c'd.

First and last release both in 2014, which is not encouraging.

**Comment by ntamas** [comment]
It seems like adopting FEAST would basically mean committing ourselves to requiring a Fortran compiler at build time; FEAST does not seem to be packaged in Debian/Ubuntu yet and it has no formula in Homebrew either (but it's at least in MacPorts). SLEPc has [Ubuntu packages](https://packages.ubuntu.com/kinetic/libslepc-real3.16), but it depends on ARPACK so I wonder whether we would gain anything by switching to SLEPc if the ultimate goal is to get rid of ARPACK.

I'd dismiss z-PARES as well due to lack of development activity.

**Comment by szhorvat** [comment]
It seems to me that all ARPACK alternatives come with significant drawbacks, and none are realistically vendorable. Replacement doesn't seem to make sense in the short term. But it may be a good to _think about_ paving the way, i.e. making the interface such that multiple eigensolver back-ends could be supported. For now I'd just consider how such an interface would look like, but I wouldn't put too much time into it.

I consider it an important feature of igraph that it can be compiled without external dependencies, i.e. that all critical dependencies are vendored. We need to keep a close relationship to our true userbase, which is researchers, not software developers.

**Comment by szhorvat** [comment]
There is also https://spectralib.org/

Spectra claims to re-implement the algorithms of ARPACK as a header-only C++ library. It depends only on Eigen. It seems to support the operations we need.

Reasons against using it:

 - The Eigen dependency makes it a heavyweight.
 - Recent Eigen uses `aligned_alloc` which is not available on macOS 10.14 and earlier. There may be workarounds, this is unclear to me. I had a bad experience where I found out that `libfive` cannot be compiled with Eigen 3.4 on macOS 10.14 because of this limitation.
 - It is written and maintained by a single person, which makes it high-risk. It is included in Debian, but not in Arch or Fedora (most Linux distros are derived from one of these three). https://repology.org/project/spectra/versions

**Comment by szhorvat** [comment]
There are more features that ARPACK provides that we rely on, and should be available in any replacement library. For example, ARPACK allows for finding the eigenvalue with the largest real part, instead of the largest magnitude. We rely on this for PageRank to find the eigenvector corresponding to the eigenvalue `1`, and _not_ `-1`, even though both may be present.

Having a common interface to multiple solvers may be problematic, as selecting the desired eigenvalue would be achieved in a different manner with each.

It is more realistic to look for libraries that implement the same algorithm as ARPACK, and are preferably written in C without heavy dependencies. Spectra would be a candidate if it didn't have the Eigen dependency.

**Comment by krlmlr** [comment]
CRAN is now coming after us because ARPACK doesn't pass the new `R CMD check` :

![image](https://github.com/user-attachments/assets/9e92957f-f661-4992-9da6-2fb08d149bd8)

Mainly this is due to the usage of "old-style character vectors":

```
gfortran-14 -fvisibility=hidden -fpic  -g -O2 -mtune=native -Wall -pedantic  -c vendor/arpack/dmout.f -o vendor/arpack/dmout.o
vendor/arpack/dmout.f:34:19:

   34 |       CHARACTER*80       LINE
      |                   1
Warning: Obsolescent feature: Old-style character length at (1)
```

Link: https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-fedora-gcc/igraph-00install.html .

I'm trying to push the current release despite the warning, let's see. We do need a solution at some point, though.

I've found https://bitbucket.org/chaoyang2013/arpack/src/master/, but no commits in the last two years. Do we patch in-place, or try to upstream?

**Comment by szhorvat** [comment]
This is an R/igraph-specific issue (and ARPACK won't be replaced), so continuing here: https://github.com/igraph/rigraph/issues/1533

**Comment by ilayn** [comment]
Hi SciPy maintainer here, 

Just out of curiosity I was searching ARPACK issues on GitHub and found this. As part of our F77 removal saga, https://github.com/scipy/scipy/issues/18566, I have rewritten ARPACK in C from scratch https://github.com/scipy/scipy/pull/22748 (not passing it through f2c but actually rewriting it) some mechanically some by rewriting the algorithm at places. The reasons why we are doing it is a subject of another discussion altogether (and hopefully better be skipped for now) however I finished the non-symmetric real and complex solvers and in a few days symmetric parts will be done.

It's neither the best code nor am I a C expert, but seemingly it works for large problems and with all compiler warning flags are turned on I don't get response from compilers. At this point, I just to emphasize that our first concern is to remove F77 code from SciPy then other auxiliary benefits of having C code. I removed a ton of Gotos but another 10 tons of Goto still exists. However the remaining Gotos are pretty much flow control and not so terrible hidden while loops as before. I also removed all SAVEd variables hence now it is thread-safe since the reverse communication is done through a C struct.

The reason why I am disrupting this issue (apologies for that) is whether there is any interest for such code and also if any experts wants to implement something while the patient is still not yet stitched up. As an example  I replaced and removed the random vector generation outside ARPACK and now is performed on Python side. This allowed us to have reproducible runs by fixing the seed outside. 

By these rewrites that you can see in the main issue linked above, I'm hoping that we can centralize these tools for all platforms and packages/languages without resorting to a Fortran compiler. 

Please feel free to ignore as it does not solve the fundamental problems of ARPACK but fixes a few bugs along the way. I can ping you again once I am done testing if you are interested.

**Comment by szhorvat** [comment]
@ilayn I am very interested in collaborating on this. Did you rewrite the original ARPACK or the updated ARPACK-NG?

I am quite overwhelmed with work now but I'd be happy to test it with igraph in a couple of weeks, and I'm interested in bringing it to a state where it works with igraph (and ideally, integrated with igraph's own random number generator).

One of the issues we encountered was that CRAN is very fussy about the quality of C code in R packages. Thus the f2c version was a no-go for igraph's R interface (and was only used by the other interfaces). For this reason, the R interface still uses the Fortran code. Getting this of this, and replacing it with C code that is the same for all igraph interfaces would be another big win in my opinion.

**Comment by ilayn** [comment]
Yes, this is based on ARPACK-NG 3.9.1, including that zero vector leading to error -9 which I handle on Python side by returning nev-many 0 eigenvalues and that many identity matrix columns as the result. 

No rush as we have about 6 months to have this in the new SciPy release 1.17 (1.16 is about to hit the streets soon)

**Comment by szhorvat** [comment]
@ilayn, would you be willing to factor out the rewrite into a standalone library, avoiding hard entanglements with SciPy?

**Comment by ilayn** [comment]
Yes of course, in fact it is already in a standalone state. All SciPy related parts are stuffed into the `arpackmodule.c` file and the rest is compiled as a static library (with linkage to BLAS/LAPACK obviously) with no python related things in it (everything in this folder https://github.com/ilayn/scipy/tree/rewrite_arpack/scipy/sparse/linalg/_eigen/arpack/ARPACK). 

Apologies for my ignorance about distributing a C code and its -isms, so probably it does not look like it, but I don't like build systems so you just compile all files together and that's pretty much it. 

If you wish to have make file for that (to discover MKL/BLIS/Accelerate etc.), no problem from my side. I just I don't know much other than via pkg-config tool settings. We use meson system for that hence it's slightly easier for me to rely upon that doing the complicated work.

**Comment by ilayn** [comment]
I probably finish up first and properly document the API changes then it should be obvious. You can see the signatures of the new interfaces in the header file, and the structs (single and double precision) used to pass the gory parameters that were contaminating the calls previously. Now you can instantiate the structs to whatever you want to do and it modifies them in-place. Hence you can even watch the progress during the reverse communication steps.

https://github.com/ilayn/scipy/blob/rewrite_arpack/scipy/sparse/linalg/_eigen/arpack/ARPACK/_arpack.h

Once I am done with it, I'll externalize the int type to allow for ILP64 systems so it would be a simple switch between 32ints and 64ints.

But probably you folks know a lot more than I do maintaining a C codebase hence I'm all ears if you want to correct/teach me.

**Comment by ilayn** [comment]
Is there any ideas about what I should call the new API and the C code? I am really out of ideas for a name and I don't wish to have a name that has ARPACK in it to avoid confusion. I think Spectra and ArnoldiMethods.jl did the smart thing by properly renaming their code. So possibly this new code should do the same. 

I'm going to start the documentation work hence I'm looking for something to replace "ARPACK" and populate with. I was about to use the wacky CARPARK name to play with the letters out of desperation but probably won't be received well :)

**Comment by szhorvat** [comment]
> to use the wacky CARPARK name to play with the letters

Well, there's CLAPACK and CBLAS, which are not entirely dissimilar. But these are not so much C rewrites as a more idiomatic C interface.

If you're looking for a prettier name, how about simply Arnoldi, with `arnoldi.h`?

**Comment by ilayn** [comment]
Yes I guess it is the obvious choice though there are like 5 arnoldis around with varying adoption. Anyways, let's call it Arnoldi as a running name. I'll start documentation and other details. Thanks

**Comment by ilayn** [comment]
I've renamed it to Arnaud (same name but a bit of dutchified version) and added a slightly modernized structure here https://github.com/scipy/scipy/pull/23274

**Comment by szhorvat** [comment]
@ilayn We are planning an major release of igraph within a month. After that we won't be able to break API for some time.

I'd like to fully move igraph to using Arnaud, but I don't have the time to do this work before our big release.

My question to you: Are there any changes in Arnaud, compared to ARPACK, that would force an API-break in igraph? If yes, I'd like to make that API change on our side now.

igraph's ARPACK use is encapsulated in two functions,

https://igraph.org/c/html/latest/igraph-Arpack.html#igraph_arpack_rnsolve
https://igraph.org/c/html/latest/igraph-Arpack.html#igraph_arpack_rssolve

The key here is the `igraph_arpack_options_t` parameter, which sets options for ARPACK. You can see it here:

https://igraph.org/c/html/latest/igraph-Arpack.html#igraph_arpack_options_t

Do you notice anything that would absolutely NEED to be changed here once we move to Arnaud? I don't think there's anything, right? The internal `ipntr` and `iparam` are ugly, but can be adapted. Same goes for the string values of `which`: ugly, but can be adapted. If I have time, I might refactor these to an enum before our big release.

----

The `igraph_arpack_storage_t`, as most of this stuff, is from well before my time, but it seems to me that it's purpose is to be able to allocate temporary storage (such as `workl`, `workd`, etc.) and use it multiple times, without re-allocation.

Is this still relevant for Arnaud, @ilayn? Do you think it's a good feature to have?

@ntamas, I see that this feature is never used. All calls to our ARPACK interface pass `NULL` for `storage`, meaning that the temporary storage is allocated automatically. For this reason, I am tempted to remove this parameter entirely before igraph 1.0. Let me know if you agree.

----

@ilayn I have to admit that some of these questions come from time pressure from the release deadline, and simply not having the time to look at Arnaud (and recap ARPACK) in sufficient detail. If you can give any advice, it'd be much appreciated, as I hope to move to Arnaud exclusively in a future igraph version.

**Comment by ntamas** [comment]
@szhorvat Feel free to remove the feature if it is not used. igraph is not meant to be a generic ARPACK interface so it's okay if we don't expose all the features that we could.

**Comment by ilayn** [comment]
Hi @szhorvat 

Here is a link in case you need to peek into it after reading below [scipy/sparse/linalg/_eigen/arpack/arnaud](https://github.com/scipy/scipy/tree/main/scipy/sparse/linalg/_eigen/arpack/arnaud)

Let me try to emphasize some changes; 

- Arnaud is a one-to-one complete rewrite. So there are no changes in terms of the computations. We tested as much as we could and results so far agree. It is too early to call it bug free however it seems like it is surviving heavy problems. But still, there would be some problems that will not converge due to thealgorithm shortcomings. However now you can setup your favorite debugger and look for reasons, instead of pulling your hair out. 
- The things we cleaned up are the number of OP*x, or B*x or number of reorthogonalizations etc., all the timing related noise and also the now unused iparam vector. 
- We linked to BLAS/LAPACK in the way that works for us that is lower case + an underscore, `dgetrf_`, `sgemm_` but these are in a single place. If you want to modify them there is a single place that needs surgery or additions see `blaslapack_declarations.h`.
- The other change is the data type of certain switches such as `bmat` being now an `int` acting like a `bool` instead of the dreaded Fortran `character`s. 
- I see that there are some remarks in the vectors pointers with "The matrix will be resized as needed". But I am not sure what is meant there. Do you mean realloc'ing the pointer?
- The main difference is the way problem is setup and run. Previously, this was done by setting up magical indices of iparam/ipntr vector now they are all inside a single struct marked whether they are input/output. You instantiate one of those and set the relevant ones, say with designated initializers, and bring sufficient memory and it is pretty much the same. This causes an API change and that I cannot judge whether it is acceptable for igraph. I tried to give a simple example in the readme. 
- You might bite the bullet and accept old style calling and convert them into the arnaud struct as a compatibility layer for a version or two as we do it in scipy with deprecations warnings and then remove it. But again, not sure if this is acceptable. 
- One internal change that might come at some point is the 64bit BLAS/LAPACK switch that will convert all `int`s into ARNAUD_INT such that folks with 64bit LAPACK can build by providing an environment variable. It won't change anything if you don't do anything but just a heads up in case you suddenly see the code is using strange int types.
- Overall, if you see something really wonky please let us know. Though we take pride in providing a scientific Python package to millions, we are not really C experts developing C libraries. This is kind of a desperate hail-mary attempt to get rid of the F77 mess we have been in (I now have PROPACK and ODEPACK almost completed hence soon we will be done with https://github.com/scipy/scipy/issues/18566 which grants a full night out getting shitfaced when finished heh)

I think the Makefile works, at least it installs properly. So if you already have a working reverse communication setup, here are the working changes that would require your attention, 

 1. As mentioned above, instead of the iparam ipntr stuff instantiate a struct
 2. You will not receive ido = -1 anymore which is removed but you will receive ido=4 and ido=5. 
 3. `ido=4` means fill in `resid` with `n` random doubles (or floats if single precision) with your favorite PRNG. If you need a quick one to save time I'll paste what I wrote for PROPACK. You need to seed it with 4 64bit `uints` from `stdint.h`
 4. `ido=5` means make a single OP*x regardless of `mode` value. As you might know this is not the case with `ido=1` where you need to take `mode` into account. 

That's pretty much it. I'd say give yourself an hour or so and try a 10x10 example outside of igraph. If it works, that's pretty much it. Otherwise let me know where you get stuck and I'll try to help as much as I can. The compilation should be warning free as I turned on as many flags as I could find but let me know if something complaining.

Here is the PRNG I am using inside PROPACK translation (`s` is the 4-many, nonzero entry seed array pointer). 

```C
// ==================================================================================
// Random 64bit integer generator for random floating point number purposes; based on
// https://prng.di.unimi.it/xoshiro256plus.c

static uint64_t rol64(uint64_t x, int k) {
    return (x << k) | (x >> (64 - k));
}

static uint64_t xoshiro256p(uint64_t* s) {
    uint64_t const result = s[0] + s[3];
    uint64_t const t = s[1] << 17;

    s[2] ^= s[0];
    s[3] ^= s[1];
    s[1] ^= s[2];
    s[0] ^= s[3];

    s[2] ^= t;
    s[3] = rol64(s[3], 45);

    return result;
}

double random_double(uint64_t* state) {
    union {
        uint64_t u;
        double d;
    } x;

    uint64_t r = xoshiro256p(state);
    r >>= 11;  // top 53 bits
    x.u = ((uint64_t)1023 << 52) | r;
    return x.d - 1.0;
}

float random_float(uint64_t* state) {
    union {
        uint32_t u;
        float f;
    } x;

    uint32_t r = (uint32_t)(xoshiro256p(state) >> 32);  // top 32 bits, waste half of the bits
    r >>= 9;  // keep top 23 bits
    x.u = (127 << 23) | r;

    return x.f - 1.0f;
}

// ==================================================================================
```

**Comment by ilayn** [comment]
> Is this still relevant for Arnaud, @ilayn? Do you think it's a good feature to have?

I'd say yes to this one. Because you allocate once and it is passed back and forth and Arnaud does not allocate free anything. So you can do whatever you want with that memory. But they don't need to be inside a struct you can allocate them independently or allocate once a big one and slice it with pointers. Currently Arnaud does not do any memory management and expects large enough memory associated with the input pointers..

**Comment by ilayn** [comment]
Hi all, apologies for reigniting this issue. The translated version has been live in SciPy for at least two versions and nobody is running after us with pitchforks yet. That gives me confidence that it is functioning sufficiently well out in the wild. The builds are clean and as far as we can tell, warning-free for all mainstream compilers.

Now I have a bit of time to spare and I can use it to swap out the ARPACK with the translated C version as a PR for you to investigate but I have a few questions about how much of a surgery you folks have appetite for. 

A couple of things I've noticed; 

 1. `IGRAPH_USE_INTERNAL_ARPACK` and the `ARPACK_IS_VENDORED` mechanism does not apply if I swap with the C implementation.
 2. `igraph_arpack_options_t` won't be necessary as the C implementation comes with its dedicated struct. 
 3. These seven (I think) the seven callers (`src/misc/embedding.c`, `src/linalg/eigen.c`, `src/core/sparsemat.c`, `src/community/leading_eigenvector.c`, `src/centrality/eigenvector.c`, `src/centrality/hub_authority.c`, `src/centrality/pagerank.c`) needs to be touched to populate Arnaud's struct.
 4. The ARPACK branches from `tools/lapack/getlapack.sh` (lines handling ${arpackdir}, the ARPACK-specific ordering) will be gone and `tools/arpack-sed.txt` will be deleted. Also ARPACK files from `tools/lapack/mt.patch` is going to be pruned.
 5. ARPACK code gets removed from the `vendor/lapack` dir
 6. Some more minor surgeries to `CMakeLists.txt` files. 
 7. I can make the "use system ARPACK" option a no-op so at least it does not break anything. But not sure if this is acceptable.

Because I am a total stranger to this library usage from user point-of-view, not sure what constitutes as ABI/API breakage from your side hence just wanted to ask your opinion. Totally fine if you don't have any desire to touch things.

**Comment by szhorvat** [comment]
> Now I have a bit of time to spare and I can use it to swap out the ARPACK with the translated C version as a PR for you to investigate but I have a few questions about how much of a surgery you folks have appetite for.

I will need to discuss this further with other igraph devs, but personally I believe that the best way forward is to swap ARPACK for Arnaud. The primary reason why this wasn't moving forward is that I simply did not have the time to spare.

If you can assist with this, that would be a huge help.

I'll get back to you with detailed answers shortly.
